### ISSUE TITLE
invalid NodeType errors for some parsers after upgrade to 0.24.2

### ISSUE THREAD

**Issue by alexaandru** [tree-sitter/tree-sitter#3759]
### Problem

I maintain a Go library of tree sitter parsers (https://github.com/alexaandru/go-sitter-forest). I updated it yesterday from tree-sitter 0.23.2 to 0.24.2 and rebuilt all parsers. When I tried to use "some of them" (about 130 of them, my text editor acts as "integration tests" for now, about 10 of them threw invalid NodeType errors, whereas before the upgrade they were working fine. I can collect latter all the examples, but here it is just one of them: 

For tree-sitter-sql I get `invalid NodeType 'keyword_max' at ...`. When I look in parser.c (generated with 0.24.2) I do not see a `sym_keyword_max` I only see `sym_keyword_maxvalue` so the max one has disappeared for some reason. If I look in the `parser.c` that project includes at https://raw.githubusercontent.com/DerekStride/tree-sitter-sql/refs/heads/gh-pages/src/parser.c you can see it right there `sym_keyword_max = 70`. That file was generated 2mo ago, so was definitely not compiled with 0.24.x.

The grammar is fine, you can see it present there still: https://github.com/DerekStride/tree-sitter-sql/blob/main/grammar.js#L133

The only change between their stored version of `parser.c` and my stored version of their `parser.c` is that mine was generated with the latest tree-sitter-cli (whatever was the latest ~16 hours ago): https://github.com/alexaandru/go-sitter-forest/blob/main/sql/parser.c  I've used this query FWIW on the tree: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/sql/highlights.scm

<details><summary>the list of all parsers that failed with the same error</summary>

- apex: invalid NodeType 'default' at line 389 column 0
- astro: invalid NodeType 'entity' at line 115 column 0
- bitbake: invalid NodeType 'INHERIT' at line 105 column 0
- dart: invalid NodeType 'dynamic' at line 245 column 0
- scala: invalid NodeType 'symbol_literal' at line 160 column 0
- sql: invalid NodeType 'keyword_max' at line 155 column 0
- starlark: invalid NodeType 'await' at line 253 column 0
- verilog: invalid NodeType 'config' at line 13 column 0
- xml: invalid NodeType 'INCLUDE' at line 149 column 0
</details>

If there are any other details I can provide, happy to do so.

Cheers!

### Steps to reproduce

git clone --depth=1 https://github.com/DerekStride/tree-sitter-sql
cd tree-sitter-sql
npx tree-sitter --version # tree-sitter 0.24.2 (0b167b0782a74e3702fb77cbef1486c6c01c24c8)
npx tree-sitter generate
grep keyword_max grammar.js src/parser.c|sort -u
# you will see it defined in `grammar.js` but NOT in `parser.c`

### Expected behavior

Expected to see `sym_keyword_maxvalue = 123` (some int, not 123 specifically) in generated `parser.c`.

### Tree-sitter version (tree-sitter --version)

0.24.2

### Operating system/version

Fedora Linux 40 (Workstation Edition)

**Comment by sogaiu** [comment]
When I used 0.23.2 to generate `parser.c` and friends, I got the same results as with 0.24.2, specifically for the DerekStride version of tree-sitter-sql.

Both were missing `keyword_max`.

---

I did see the difference described when I used 0.23.0 (compared with 0.24.2).  That is, 0.23.0 had `keyword_max` while 0.24.2 did not.

Don't know if there's some pilot error on my part, but FWIW (^^;

**Comment by alexaandru** [comment]
Yeah, just checked, https://github.com/DerekStride/tree-sitter-sql/blob/main/package.json#L23 it appears that the parser was built with "0.22.6", at least according to: https://github.com/DerekStride/tree-sitter-sql/blob/main/package-lock.json#L3260-L3269

My bad, I stand corrected. I think the last time I rebuilt all 400+ parsers was after the upgrade to `0.23.0`. Which explains why "it worked last time". I'll update the description, thank you! :-)

**Comment by amaanq** [comment]
This is intended, as per #3726. Unused rules in a grammar are filtered out of the final parser.c since it fixed several crashes later on. It is in fact a breaking change, and is marked as such in the release notes.
